File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/component/ComponentDependency.java
package ru.tinkoff.kora.kora.app.annotation.processor.component;

import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.CodeBlock;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.kora.app.annotation.processor.GraphResolutionHelper;
import ru.tinkoff.kora.kora.app.annotation.processor.ProcessingContext;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclaration;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclarations;

import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import java.util.Objects;

public sealed interface ComponentDependency {

    DependencyClaim claim();

    default CodeBlock write(ProcessingContext ctx, ClassName graphTypeName, ComponentDeclarations componentDeclarations, ResolvedComponents resolvedComponents) {
        return switch (this) {
            case AllOfDependency(var claim) -> {
                var dependencyDeclarations = GraphResolutionHelper.findDependencyDeclarations(ctx, componentDeclarations, claim);
                var codeBlock = CodeBlock.builder().add("$T.of(", CommonClassNames.all);
                var dependencies = GraphResolutionHelper.findDependenciesForAllOf(ctx, claim, dependencyDeclarations, resolvedComponents);
                for (int i = 0; i < dependencies.size(); i++) {
                    var dependency = dependencies.get(i);
                    if (i == 0) {
                        codeBlock.indent().add("\n");
                    }
                    codeBlock.add(dependency.write(ctx, graphTypeName, componentDeclarations, resolvedComponents));
                    if (i == dependencies.size() - 1) {
                        codeBlock.unindent();
                    } else {
                        codeBlock.add(",");
                    }
                    codeBlock.add("\n");
                }

                yield codeBlock.add("  )").build();
            }
            case NullDependency(var claim) -> switch (claim.claimType()) {
                case ONE_NULLABLE -> CodeBlock.of("($T) null", claim.type());
                case NULLABLE_VALUE_OF -> CodeBlock.of("($T<$T>) null", CommonClassNames.valueOf, claim.type());
                case NULLABLE_PROMISE_OF -> CodeBlock.of("($T<$T>) null", CommonClassNames.promiseOf, claim.type());
                default -> throw new IllegalArgumentException(claim.claimType().toString());
            };
            case PromisedProxyParameterDependency(_, var claim) -> {
                var declarations = GraphResolutionHelper.findDependencyDeclarations(ctx, componentDeclarations, claim);
                if (declarations.size() != 1) {
                    throw new IllegalStateException();
                }
                var dependency = Objects.requireNonNull(resolvedComponents.getByDeclaration(declarations.getFirst()));
                yield CodeBlock.of("g.promiseOf($T.$N.$N)", graphTypeName, dependency.holderName(), dependency.fieldName());
            }
            case PromiseOfDependency(var claim, var delegate) when delegate instanceof WrappedTargetDependency ->
                CodeBlock.of("g.promiseOf($T.$N.$N).map($T::value).map(v -> ($T) v)", graphTypeName, delegate.component().holderName(), delegate.component().fieldName(), CommonClassNames.wrapped, claim.type());
            case PromiseOfDependency(var claim, var delegate) ->
                CodeBlock.of("g.promiseOf($T.$N.$N).map(v -> ($T) v)", graphTypeName, delegate.component().holderName(), delegate.component().fieldName(), claim.type());
            case TargetDependency(var _, var component) -> CodeBlock.of("g.get($T.$N.$N)", graphTypeName, component.holderName(), component.fieldName());
            case TypeOfDependency(var claim) -> TypeOfDependency.buildTypeRef(ctx.types, claim.type());
            case ValueOfDependency(var claim, var delegate) when delegate instanceof WrappedTargetDependency ->
                CodeBlock.of("g.valueOf($T.$N.$N).map($T::value).map(v -> ($T) v)", graphTypeName, delegate.component().holderName(), delegate.component().fieldName(), CommonClassNames.wrapped, claim.type());
            case ValueOfDependency(var claim, var delegate) ->
                CodeBlock.of("g.valueOf($T.$N.$N).map(v -> ($T) v)", graphTypeName, delegate.component().holderName(), delegate.component().fieldName(), claim.type());
            case WrappedTargetDependency(var _, var component) -> CodeBlock.of("g.get($T.$N.$N).value()", graphTypeName, component.holderName(), component.fieldName());
        };
    }

    sealed interface SingleDependency extends ComponentDependency {
        ResolvedComponent component();
    }

    record TargetDependency(DependencyClaim claim, ResolvedComponent component) implements SingleDependency {

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("TargetDependency[");
            sb.append("claim=").append(claim);
            sb.append(", index=").append(component.index());
            sb.append(", fieldName=").append(component.fieldName());
            sb.append(", holder=").append(component.holderName());
            sb.append(", declaration=").append(component.declaration());
            if (component.templateParams() != null && !component.templateParams().isEmpty()) {
                sb.append(", templateParams=").append(component.templateParams());
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record WrappedTargetDependency(DependencyClaim claim, ResolvedComponent component) implements SingleDependency {

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("WrappedTargetDependency[");
            sb.append("claim=").append(claim);
            sb.append(", index=").append(component.index());
            sb.append(", fieldName=").append(component.fieldName());
            sb.append(", holder=").append(component.holderName());
            sb.append(", declaration=").append(component.declaration());
            if (component.templateParams() != null && !component.templateParams().isEmpty()) {
                sb.append(", templateParams=").append(component.templateParams());
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record NullDependency(DependencyClaim claim) implements ComponentDependency {
    }

    record ValueOfDependency(DependencyClaim claim, SingleDependency delegate) implements SingleDependency {
        @Override
        public ResolvedComponent component() {
            return delegate.component();
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("ValueOfDependency[");
            sb.append("claim=").append(claim);

            ResolvedComponent resolvedComponent = component();
            sb.append(", index=").append(resolvedComponent.index());
            sb.append(", fieldName=").append(resolvedComponent.fieldName());
            sb.append(", holder=").append(resolvedComponent.holderName());
            sb.append(", declaration=").append(resolvedComponent.declaration());
            if (resolvedComponent.templateParams() != null && !resolvedComponent.templateParams().isEmpty()) {
                sb.append(", templateParams=").append(resolvedComponent.templateParams());
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record PromiseOfDependency(DependencyClaim claim, SingleDependency delegate) implements SingleDependency {
        @Override
        public ResolvedComponent component() {
            return delegate.component();
        }

        @Override
        public String toString() {
            final StringBuilder sb = new StringBuilder("PromiseOfDependency[");
            sb.append("claim=").append(claim);

            ResolvedComponent resolvedComponent = component();
            sb.append(", index=").append(resolvedComponent.index());
            sb.append(", fieldName=").append(resolvedComponent.fieldName());
            sb.append(", holder=").append(resolvedComponent.holderName());
            sb.append(", declaration=").append(resolvedComponent.declaration());
            if (resolvedComponent.templateParams() != null && !resolvedComponent.templateParams().isEmpty()) {
                sb.append(", templateParams=").append(resolvedComponent.templateParams());
            }
            sb.append(']');
            return sb.toString();
        }
    }

    record TypeOfDependency(DependencyClaim claim) implements SingleDependency {
        private static CodeBlock buildTypeRef(Types types, TypeMirror typeRef) {
            if (typeRef instanceof DeclaredType) {
                var b = CodeBlock.builder();
                var typeArguments = ((DeclaredType) typeRef).getTypeArguments();

                if (typeArguments.isEmpty()) {
                    b.add("$T.of($T.class)", CommonClassNames.typeRef, types.erasure(typeRef));
                } else {
                    b.add("$T.<$T>of($T.class", CommonClassNames.typeRef, typeRef, types.erasure(typeRef));
                    for (var typeArgument : typeArguments) {
                        b.add("$>,\n$L$<", buildTypeRef(types, typeArgument));
                    }
                    b.add("\n)");
                }
                return b.build();
            } else {
                return CodeBlock.of("$T.of($T.class)", CommonClassNames.typeRef, typeRef);
            }
        }

        @Override
        public ResolvedComponent component() {
            return null;
        }
    }


    // AllOf dependencies has no resolved declaration: we will resolve them after graph building

    record AllOfDependency(DependencyClaim claim) implements ComponentDependency {
    }

    record PromisedProxyParameterDependency(ComponentDeclaration declaration, DependencyClaim claim) implements ComponentDependency {
        @Override
        public String toString() {
            return "PromisedProxyParameterDependency[claim=" + claim + ", declaration=" + declaration + ']';
        }
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/component/ComponentDependency.kt
package ru.tinkoff.kora.kora.app.ksp.component

import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSTypeAlias
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import com.squareup.kotlinpoet.ksp.toTypeParameterResolver
import ru.tinkoff.kora.kora.app.ksp.GraphResolutionHelper
import ru.tinkoff.kora.kora.app.ksp.ProcessingContext
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclarations
import ru.tinkoff.kora.ksp.common.CommonClassNames


sealed interface ComponentDependency {
    val claim: DependencyClaim

    fun write(ctx: ProcessingContext, typeToDeclarations: ComponentDeclarations, resolvedComponents: ResolvedComponents): CodeBlock = when (this) {
        is AllOfDependency -> {
            val codeBlock = CodeBlock.builder().add("%T.of(", CommonClassNames.all)
            val dependencyDeclarations = GraphResolutionHelper.findDependencyDeclarations(ctx, typeToDeclarations, claim);
            val dependencies = GraphResolutionHelper.findDependenciesForAllOf(ctx, claim, dependencyDeclarations, resolvedComponents)
            for ((i, dependency) in dependencies.withIndex()) {
                if (i == 0) {
                    codeBlock.indent().add("\n")
                }
                codeBlock.add(dependency.write(ctx, typeToDeclarations, resolvedComponents))
                if (i == dependencies.size - 1) {
                    codeBlock.unindent()
                } else {
                    codeBlock.add(", ")
                }
                codeBlock.add("\n")
            }
            codeBlock.add(")").build()
        }

        is NullDependency -> when (claim.claimType) {
            DependencyClaim.DependencyClaimType.NULLABLE_ONE -> CodeBlock.of("null as %T", claim.type.toTypeName().copy(true))
            DependencyClaim.DependencyClaimType.NULLABLE_VALUE_OF -> CodeBlock.of("null as %T", CommonClassNames.valueOf.parameterizedBy(claim.type.toTypeName()).copy(true))
            DependencyClaim.DependencyClaimType.NULLABLE_PROMISE_OF -> CodeBlock.of("null as %T", CommonClassNames.promiseOf.parameterizedBy(claim.type.toTypeName()).copy(true))
            else -> throw IllegalArgumentException(claim.claimType.toString())
        }


        is PromisedProxyParameterDependency -> {
            val declarations = GraphResolutionHelper.findDependencyDeclarations(ctx, typeToDeclarations, claim)
            if (declarations.size != 1) {
                throw IllegalStateException();
            }
            val dependency = resolvedComponents.getByDeclaration(declarations.first())!!
            CodeBlock.of("it.promiseOf(self.%N.%N)", dependency.holderName, dependency.fieldName)
        }

        is PromiseOfDependency -> {
            if (delegate is NullDependency) {
                CodeBlock.of("%T.promiseOfNull()", CommonClassNames.promiseOf)
            } else {
                val component = delegate.component!!
                if (delegate is WrappedTargetDependency) {
                    CodeBlock.of("it.promiseOf(%N.%N).map { it.value() }.map { it as %T }", component.holderName, component.fieldName, claim.type.toTypeName())
                } else {
                    CodeBlock.of("it.promiseOf(%N.%N).map { it as %T }", component.holderName, component.fieldName, claim.type.toTypeName())
                }
            }
        }

        is TargetDependency -> CodeBlock.of("it.get(%N.%N)", component.holderName, component.fieldName)
        is ValueOfDependency -> {
            if (delegate is NullDependency) {
                CodeBlock.of("%T.valueOfNull()", CommonClassNames.valueOf)
            } else {
                val component = delegate.component!!
                if (delegate is WrappedTargetDependency) {
                    CodeBlock.of("it.valueOf(%N.%N).map { it.value() }.map { it as %T }", component.holderName, component.fieldName, claim.type.toTypeName())
                } else {
                    CodeBlock.of("it.valueOf(%N.%N).map { it as %T }", component.holderName, component.fieldName, claim.type.toTypeName())
                }
            }
        }

        is WrappedTargetDependency -> CodeBlock.of("it.get(%N.%N).value()", component.holderName, component.fieldName)
        is TypeOfDependency -> buildTypeRef(claim.type)
    }

    sealed interface SingleDependency : ComponentDependency {
        val component: ResolvedComponent?
    }

    data class TargetDependency(override val claim: DependencyClaim, override val component: ResolvedComponent) : SingleDependency

    data class WrappedTargetDependency(override val claim: DependencyClaim, override val component: ResolvedComponent) : SingleDependency

    data class NullDependency(override val claim: DependencyClaim) : SingleDependency {
        override val component = null
    }


    data class ValueOfDependency(override val claim: DependencyClaim, val delegate: SingleDependency) : SingleDependency {
        override val component
            get() = delegate.component
    }

    data class PromiseOfDependency(override val claim: DependencyClaim, val delegate: SingleDependency) : SingleDependency {
        override val component
            get() = delegate.component
    }

    data class TypeOfDependency(override val claim: DependencyClaim) : ComponentDependency {
        fun buildTypeRef(typeRef: KSType): CodeBlock {
            val typeParameterResolver = typeRef.declaration.typeParameters.toTypeParameterResolver()
            var declaration = typeRef.declaration
            if (declaration is KSTypeAlias) {
                declaration = declaration.type.resolve().declaration
            }
            if (declaration is KSClassDeclaration) {
                val b = CodeBlock.builder()
                val typeArguments = typeRef.arguments

                if (typeArguments.isEmpty()) {
                    b.add("%T.of(%T::class.java)", CommonClassNames.typeRef, declaration.toClassName())
                } else {
                    b.add("%T.of(%T::class.java", CommonClassNames.typeRef, declaration.toClassName())
                    for (typeArgument in typeArguments) {
                        b.add(",\n%L", buildTypeRef(typeArgument.type!!.resolve()))
                    }
                    b.add("\n)")
                }
                return b.build()
            } else {
                return CodeBlock.of("%T.of(%T::class.java)", CommonClassNames.typeRef, typeRef.toTypeName(typeParameterResolver))
            }
        }
    }

    data class AllOfDependency(override val claim: DependencyClaim) : ComponentDependency

    data class PromisedProxyParameterDependency(val declaration: ComponentDeclaration, override val claim: DependencyClaim) : ComponentDependency

}
